modules/core/keeper/msg_server_test.go

```diff
@@ -827,7 +831,74 @@ func (suite *KeeperTestSuite) TestUpdateClientParams() {
 }
 
 // TestIBCSoftwareUpgrade tests the IBCSoftwareUpgrade rpc handler
-func (*KeeperTestSuite) TestIBCSoftwareUpgrade() {
+func (suite *KeeperTestSuite) TestIBCSoftwareUpgrade() {
+	var (
+		expError error
+		msg      *clienttypes.MsgIBCSoftwareUpgrade
+	)
+	testCases := []struct {
+		name     string
+		malleate func()
+		expPass  bool
+	}{
+		{
+			"success: valid authority and client upgrade",
+			func() {},
+			true,
+		},
+		{
+			"failure: invalid authority address",
+			func() {
+				msg.Authority = suite.chainA.SenderAccount.GetAddress().String()
+				expError = ibcerrors.ErrUnauthorized
+			},
+			false,
+		},
+		{
+			"failure: invalid clientState",
+			func() {
+				msg.UpgradedClientState = nil
+				expError = clienttypes.ErrInvalidClientType
+			},
+			false,
+		},
+		{
+			"failure: failed to schedule client upgrade",
+			func() {
+				msg.Plan.Height = 0
+				expError = sdkerrors.ErrInvalidRequest
+			},
+			false,
+		},
+	}
+
+	for _, tc := range testCases {
+		tc := tc
+		suite.Run(tc.name, func() {
+			suite.SetupTest()
+			validAuthority := suite.chainA.App.GetIBCKeeper().GetAuthority()
+			var err error
+			msg, err = clienttypes.NewMsgIBCSoftwareUpgrade(
+				validAuthority,
+				upgradetypes.Plan{
+					Name:   "upgrade IBC clients",
+					Height: 1000,
+				},
+				ibctm.NewClientState(suite.chainA.ChainID, ibctesting.DefaultTrustLevel, ibctesting.TrustingPeriod, ibctesting.UnbondingPeriod, ibctesting.MaxClockDrift, clienttypes.NewHeight(1, 10), commitmenttypes.GetSDKSpecs(), ibctesting.UpgradePath),
+			)
+
+			suite.Require().NoError(err)
+
+			tc.malleate()
+
+			_, err = keeper.Keeper.IBCSoftwareUpgrade(*suite.chainA.App.GetIBCKeeper(), suite.chainA.GetContext(), msg)
+			if tc.expPass {
+				suite.Require().NoError(err)
```

Comment:
Inside `expPass` we can also check that there exists an upgrade plan in the `x/upgrade` store and upgraded client state in the `upgradePath`